### ISSUE TITLE
🚀 Feature: Switch scaffolder actions to use zod to define input and output

### ISSUE THREAD

**Issue by Rugvip** [backstage/backstage#15439]
### 🔖 Feature description

Scaffolder actions currently use a manually defined JSON schema and separate TypeScript type for their input and output. Let's switch that over to use [zod-to-json-schema](https://www.npmjs.com/package/zod-to-json-schema) instead. The schema is verbose, and the TypeScript types frequently do not line up with the schema.

Quick little comparison between the two approaches, note that this isn't necessarily the API we'd want to end up with:

```ts
return createTemplateAction<{ url: string; targetPath?: string }>({
  id: 'fetch:plain',
  schema: {
    input: {
      type: 'object',
      required: ['url'],
      properties: {
        url: {
          title: 'Fetch URL',
          description:
            'Relative path or absolute URL pointing to the directory tree to fetch',
          type: 'string',
        },
        targetPath: {
          title: 'Target Path',
          description:
            'Target path within the working directory to download the contents to.',
          type: 'string',
        },
      },
    },
  },
  ...
})   
```

vs

```ts
return createZodTemplateAction({
  id: 'fetch:plain',
  schema: z.object({
    url: z.string().min(1).describe('Relative path or absolute URL pointing to the directory tree to fetch'),
    targetPath: z.string().optional().describe('Target path within the working directory to download the contents to.'),
  }),
  ...
})
```

### 🎤 Context

This is a more focused version of #9555, see that issue for more details. We have now used [zod-to-json-schema](https://www.npmjs.com/package/zod-to-json-schema) for both permissions and scaffolder field extensions, and it's a pretty sweet utility for compile-time and runtime checking along with schema generation. The schema generation is important when you want to be able to serialize the checks for things like remote validation or documentation.

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

**Comment by zcason** [comment]
I can take this!

**Comment by zcason** [comment]
Here's a screenshot of what I was able to come up with:
![scaffolderAction](https://user-images.githubusercontent.com/62040057/210284200-640fd7e1-92a4-40ef-84aa-f89ee9a965c5.png)

 Currently, I have a new function `convertZodToJson` that converts and parses out the expected json schema (I will place in a separate fill later). This seems to be right logically but there's minor type issue that occurring. The `input` key invokes an error `Type 'JSONSchema7Type | undefined' is not assignable to type 'Schema | undefined'`. 
 
 I've attempted to add `JSONSchema7Type` to `actions/types.ts` (ex. `input?: Schema | JSONSchema7Type;`) but receive an error `Cannot find name 'JSONSchema7Type'`.
 
 Any suggestions on a path forward?
